---
title: "Tidy Transcriptomics For Single-Cell RNA Sequencing Analyses"
author:
  - Stefano Mangiola, Walter and Eliza Hall Institute^[<mangiola.s at wehi.edu.au>]
  - Maria Doyle, Peter MacCallum Cancer Centre^[<maria.doyle at petermac.org>]
output: rmarkdown::html_vignette
bibliography: "`r file.path(system.file(package='bioc2022tidytranscriptomics', 'vignettes'), 'tidytranscriptomics.bib')`"
vignette: >
  %\VignetteIndexEntry{Tidy Transcriptomics For Single-Cell RNA Sequencing Analyses}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Instructors

*Dr. Stefano Mangiola* is currently a Postdoctoral researcher in the laboratory of Prof. Tony Papenfuss at the Walter and Eliza Hall Institute in Melbourne, Australia. His background spans from biotechnology to bioinformatics and biostatistics. His research focuses on prostate and breast tumour microenvironment, the development of statistical models for the analysis of RNA sequencing data, and data analysis and visualisation interfaces.

*Dr. Maria Doyle* is the Application and Training Specialist for Research Computing at the Peter MacCallum Cancer Centre in Melbourne, Australia. She has a PhD in Molecular Biology and currently works in bioinformatics and data science education and training. She is passionate about supporting researchers, reproducible research, open source and tidy data.

## Workshop goals and objectives

* To approach single-cell data representation and analysis through a tidy data paradigm, integrating tidyverse with tidySingleCellExperiment.
* Compare SingleCellExperiment and tidy representation  
* Apply tidy functions to SingleCellExperiment objects  
* Reproduce a real-world case study that showcases the power of tidy single-cell methods

### What you will learn

-   Basic `tidy` operations possible with `tidySingleCellExperiment`
-   The differences between `SingleCellExperiment` representation and `tidy` representation
-   How to interface `SingleCellExperiment` with tidy manipulation and visualisation
-   A real-world case study that will showcase the power of `tidy` single-cell methods compared with base/ad-hoc methods

### What you will *not* learn

-   The molecular technology of single-cell sequencing
-   The fundamentals of single-cell data analysis
-   The fundamentals of tidy data analysis

This workshop will demonstrate a real-world example of using tidy transcriptomics packages, such as tidySingleCellExperiment and tidybulk, to perform a single cell analysis. This workshop is not a step-by-step introduction in how to perform single-cell analysis. For an overview of single-cell analysis steps performed in a tidy way please see the [ISMB2021 workshop](https://tidytranscriptomics-workshops.github.io/ismb2021_tidytranscriptomics/articles/tidytranscriptomics.html).

## Slides

<iframe 
    src="https://docs.google.com/gview?url=https://raw.githubusercontent.com/tidytranscriptomics-workshops/bioc2022_tidytranscriptomics/master/inst/bioc2022_tidytranscriptomics.pdf&embedded=true" 
    scrolling="yes" 
    style="width:100%; height:600px;" 
    frameborder="0">
</iframe>


## Getting started

### Cloud

Easiest way to run this material. We will use the Orchestra Cloud platform during the BioC2022 workshop.

-   Using the URL provided launch the workshop called "BioC2022: Tidy Transcriptomics For Single-Cell RNA Sequencing Analyses" **There are several tidy transcriptomics workshops. Be sure to select the BioC2022 one**.
-   Open `tidytranscriptomics_case_study.Rmd` in `bioc2022_tidytranscriptomcs/vignettes` folder

### Local

We will use the Orchestra Cloud platform during the BioC2022 workshop and this method is available if you want to run the material after the workshop. If you want to install on your own computer, see instructions [here](https://tidytranscriptomics-workshops.github.io/bioc2022_tidytranscriptomics/index.html#workshop-package-installation).

Alternatively, you can view the material at the workshop webpage [here](https://tidytranscriptomics-workshops.github.io/bioc2022_tidytranscriptomics/articles/tidytranscriptomics_case_study.html).


## Introduction to tidySingleCellExperiment

```{r message = FALSE}
# Load packages
library(SingleCellExperiment)
library(ggplot2)
library(plotly)
library(dplyr)
library(colorspace)
library(dittoSeq)
```

SingleCellExperiment is a very popular analysis toolkit for single cell RNA sequencing data [@butler2018integrating; @stuart2019comprehensive].

Here we load single-cell data in SingleCellExperiment object format. This data is peripheral blood mononuclear cells (PBMCs) from metastatic breast cancer patients.


```{r}
# load single cell RNA sequencing data
sce_obj <- bioc2022tidytranscriptomics::sce_obj

# take a look
sce_obj
```

tidySingleCellExperiment provides a bridge between the SingleCellExperiment single-cell package and the tidyverse [@wickham2019welcome]. It creates an invisible layer that enables viewing the SingleCellExperiment object as a tidyverse tibble, and provides SingleCellExperiment-compatible *dplyr*, *tidyr*, *ggplot* and *plotly* functions.

If we load the *tidySingleCellExperiment* package and then view the single cell data, it now displays as a tibble.

```{r message = FALSE}
library(tidySingleCellExperiment)

sce_obj
```

It can be interacted with using [SingleCellExperiment commands](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html) such as `assay`.

```{r}
Assays(sce_obj)
```

We can also interact with our object as we do with any tidyverse tibble.

### Tidyverse commands

We can use tidyverse commands, such as `filter`, `select` and `mutate` to explore the tidySingleCellExperiment object. Some examples are shown below and more can be seen at the tidySingleCellExperiment website [here](https://stemangiola.github.io/tidySingleCellExperiment/articles/introduction.html#tidyverse-commands-1).

We can use `filter` to choose rows, for example, to see just the rows for the cells in G1 cell-cycle stage. Check if have groups or ident present.

```{r}
sce_obj |> filter(Phase == "G1")
```

We can use `select` to choose columns, for example, to see the sample, cell, total cellular RNA

```{r}
sce_obj |> select(.cell, nCount_RNA, Phase)
```

We can use `mutate` to create a column. For example, we could create a new `ident_l` column that contains a lower-case version of `ident`.

```{r}
sce_obj |>
  mutate(Phase_l = tolower(Phase)) |>
  # select columns to view
  select(Phase, Phase_l)
```

We can use tidyverse commands to polish an annotation column. We will extract the sample, and group information from the file name column into separate columns.

```{r}
# First take a look at the file column
sce_obj |> select(file)
```

```{r}
# Create columns for sample and group
sce_obj <- sce_obj |>
  # Extract sample and group
  extract(file, "sample", "../data/.*/([a-zA-Z0-9_-]+)/outs.+", remove = FALSE)

# Take a look
sce_obj |> select(sample)
```

We could use tidyverse `unite` to combine columns, for example to create a new column for sample id combining the sample and BCB columns.

```{r}
sce_obj <- sce_obj |> unite("sample_id", sample, BCB, remove = FALSE)

# Take a look
sce_obj |> select(sample_id, sample, BCB)
```


## Case study

### Data pre-processing

The object `sce_obj` we've been using was created as part of a study on breast cancer systemic immune response. Peripheral blood mononuclear cells have been sequenced for RNA at the single-cell level. The steps used to generate the object are summarised below.

-   `scran`, `scater`, and `DropletsUtils` packages have been used to eliminate empty droplets and dead cells. Samples were individually quality checked and cells were filtered for good gene coverage.

-   Variable features were identified using `modelGeneVar`.

-   Read counts were scaled and normalised using logNormCounts from `scuttle`.

-   Data integration was performed using `fastMNN` with default parameters.

-   PCA performed to reduce feature dimensionality.

-   Nearest-neighbor cell networks were calculated using 30 principal components.

-   2 UMAP dimensions were calculated using 30 principal components.

-   Cells with similar transcriptome profiles were grouped into clusters using Louvain clustering from `scran`.

### Analyse custom signature

The researcher analysing this dataset wanted to to identify gamma delta T cells using a gene signature from a published paper [@Pizzolato2019].

With tidySingleCellExperiment's `join_features` the counts for the genes could be viewed as columns.

```{r}

sce_obj |>
  join_features(c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide")
```

They were able to use tidySingleCellExperiment's `join_features` to select the counts for the genes in the signature, followed by tidyverse `mutate` to easily create a column containing the signature score.

```{r}

sce_obj |>
  join_features(c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide") |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  select(signature_score, everything())
```

The gamma delta T cells could then be visualised by the signature score using Bioconductor's visualisation functions.

```{r}

sce_obj |>
  join_features(
    features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide"
  ) |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  scater::plotUMAP(colour_by = "signature_score")
```

The cells could also be visualised using the popular and powerful `ggplot2` package, enabling the researcher to use ggplot functions they were familiar with, and to customise the plot with great flexibility.

```{r}

sce_obj |>
  join_features(
    features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide"
  ) |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  arrange(signature_score) |>
    
  ggplot(aes(UMAP_1, UMAP_2, color = signature_score)) +
  geom_point() +
  scale_color_distiller(palette = "Spectral") +
  bioc2022tidytranscriptomics::theme_multipanel
```

For exploratory analyses, we can select the gamma delta T cells, the red cluster on the left with high signature score. We'll filter for cells with a signature score > 0.8.

```{r}

sce_obj_gamma_delta <-
  sce_obj |>
    
  join_features(
    features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide"
  ) |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  filter(signature_score > 0.7)
```

For comparison, we show the alternative using base R and SingleCellExperiment

```{r eval=FALSE}

counts_positive <-
  GetAssayData(sce_obj, assay = "SCT")[c("CD3D", "TRDC", "TRGC1", "TRGC2"), ] |>
  colSums() |>
  scales::rescale(to = c(0, 1))

counts_negative <-
  GetAssayData(sce_obj, assay = "SCT")[c("CD8A", "CD8B"), ] |>
  colSums() |>
  scales::rescale(to = c(0, 1))

sce_obj$signature_score <- counts_positive - counts_negative

FeaturePlot(sce_obj, features = "signature_score")

sce_obj |>
  subset(signature_score > 0.8)
```

It is then possible to perform analyses on these gamma delta T cells by simply chaining further commands, such as below.

```{r eval = FALSE}
library(batchelor)
library(igraph)
library(scran)
library(magrittr)

sce_obj_gamma_delta = 
	sce_obj_gamma_delta |>
	
  # Reanalyse
	multiBatchNorm(batch = colData(sce_obj_gamma_delta)$sample) |> 
	fastMNN(batch = colData(sce_obj_gamma_delta)$sample) |> 
	
	# Join old information
	left_join(as_tibble(sce_obj_gamma_delta)) |> 
 
	# Dimension reduction
  runUMAP(ncomponents = 2, dimred="corrected") 
  	
  # Assign clusters to the 'colLabels'
	# of the SingleCellExperiment object
sce_obj_gamma_delta = 
	sce_obj_gamma_delta |> 
	mutate(
		cluster = 
			colLabels(sce_obj_gamma_delta) <- # from SingleCellExperiment
			sce_obj_gamma_delta %>%
  		buildSNNGraph(use.dimred="corrected") %>% # from scran - shared nearest neighbour
			cluster_walktrap() %$% # from igraph
			membership
	)
	

sce_obj_gamma_delta |> select(batch, cluster, everything())
```

It was also possible to visualise the cells as a 3D plot using plotly.
The example data used here only contains a few genes, for the sake of time and size in this demonstration, but below is how you could generate the 3 dimensions needed for 3D plot with a full dataset.

```{r eval = FALSE}
single_cell_object |>
  RunUMAP(dims = 1:30, n.components = 3L, spread = 0.5, min.dist = 0.01, n.neighbors = 10L)
```

We'll demonstrate creating a 3D plot using some data that has 3 UMAP dimensions.

```{r umap plot 2, message = FALSE, warning = FALSE}
pbmc <- bioc2022tidytranscriptomics::sce_obj_UMAP3

pbmc |>
  plot_ly(
    x = ~`UMAP_1`,
    y = ~`UMAP_2`,
    z = ~`UMAP_3`,
    color = ~cell_type,
    colors = dittoSeq::dittoColors()
  ) %>%
  add_markers(size = I(1))
```

## Exercises

1. What proportion of all cells are gamma-delta T cells? Use signature_score > 0.8 to identify gamma-delta T cells.

2. There is a cluster of cells characterised by a low RNA output (nCount_RNA < 100). Identify the cell composition (cell_type) of that cluster.

# Pseudobulk analyses

It is sometime useful to aggregate cell-wise transcript abundance into pseudobulk samples. It is possible to explore data and perform hypothesis testing with tools and data-source that we are more familiar with. For example, we can use edgeR in tidybulk to perform differential expression testing. For more details on pseudobulk analysis see [here](https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html).


### Data exploration using pseudobulk samples

To do this, we will use a helper function called `aggregate_cells`, available in this workshop package, to combine the single cells into groups for each cell type for each sample.

```{r warning=FALSE, message=FALSE, echo=FALSE}
library(glue)
library(tidyr)
library(tidybulk)
library(tidySummarizedExperiment)
library(purrr)
library(patchwork)
```

```{r}
pseudo_bulk <-
  sce_obj |>
  bioc2022tidytranscriptomics::aggregate_cells(c(sample, cell_type), assays = "counts") 

pseudo_bulk
```

### Tidybulk and tidySummarizedExperiment

With `tidySummarizedExperiment` and `tidybulk` is easy to stratify our dataset for iterative self-contained analyses.

# <img src="../inst/new_SE_usage-01.png" width="800px" />


To explore the grouping, we can use tidyverse `slice` to choose a row (cell_type) and `pull` to extract the values from a column. If we pull the data column we can view the SummarizedExperiment object. 

```{r}
pseudo_bulk |>
  nest(data = -cell_type)


pseudo_bulk |>
  nest(data = -cell_type) |> 
  slice(1) |>
  pull(data)
```

We can then identify differentially expressed genes for each cell type for our condition of interest, progressive versus stable metastatic breast cancer.

```{r message=FALSE, warning=FALSE}
# Differential transcription abundance
pseudo_bulk <-
  pseudo_bulk |>

  nest(data = -cell_type) |> 
  
  mutate(data = map(
    data,
    ~ .x |>
      identify_abundant(factor_of_interest = treatment) |>
      scale_abundance() |>
      test_differential_abundance(~treatment)
  ))

```


```{r message = FALSE}
pseudo_bulk <-
    
	pseudo_bulk |>
    
  # Filter out significant
  mutate(data = map(data, ~ filter(.x, FDR < 0.5))) |>
	
	# Filter cell types with no differential abundant gene-transcripts
  filter(map_int(data, ~ nrow(.x)) > 0) |>
    
  # Plot
  mutate(plot = map2(
    data, cell_type,
    ~ .x |>
      ggplot(aes(treatment, counts_scaled + 1)) +
      geom_boxplot(aes(fill = treatment)) +
      geom_jitter() +
      scale_y_log10() +
      facet_wrap(~.feature) +
      ggtitle(.y)
  ))

```

The output is a nested table with a column containing a plot for each cell type.

```{r}
pseudo_bulk
```

We'll use slice and pull again to have a look at one of the plots.

```{r}
pseudo_bulk |>
  slice(1) |>
  pull(plot)
```

We can extract all plots and plot with `wrap_plots` from the patchwork package. 

```{r, fig.height = 8}
pseudo_bulk |>
  pull(plot) |>
  wrap_plots() &
  bioc2022tidytranscriptomics::theme_multipanel
```


**Session Information**

```{r}
sessionInfo()
```

**References**
